src/asm/lexer.cpp

```diff
@@ -1849,7 +1849,13 @@ static Token yylex_NORMAL() {
 					}
 				}
 
-				if (token.type == T_(ID) && (lexerState->atLineStart || peek() == ':'))
+				// This is a "lexer hack"! We can't distinguish `ID` and `LABEL` at the parser
+				// level because we can't parse `ID` followed by `COLON`, since that's a macro
+				// invocation and enters raw mode, which lexes string arguments until EOL.
+				// Instead, we check for the exact character ':' after a non-local identifier.
+				// This means that "Label:" and "mac:" are treated as label definitions, but
+				// "Label :" and "mac :" are treated as macro invocations.
```

Comment:
```suggestion
				// This is a "lexer hack"! Here is why it is necessary:
				//
				// The parser needs to distinguish between a label definition and a macro invocation,
				// and we can't do so by having two rules starting with `T_ID`.
				//
				// Why? Well, to disambiguate between the two, the parser performs a "lookahead",
				// i.e. it makes an extra call to the lexer, to determine which rule applies.
				// *But*, this is unacceptable in this case, because macro argument *must* be
				// lexed in "raw" mode, whereas the colon following a label definition cannot be
				// recognised in "raw" mode, due to its very nature.
				//
				// This is solved by requiring that label definitions are *immediately* followed
				// by a colon, with no intervening whitespace.
				// This allows the first macro argument to start with a colon:
				// `Label:` and `makro:` are treated as label definitions (and potentially
				// cause "conflicting definition" errors), whereas `Label :` and `makro :` are
				// treated as macro invocations (and potentially cause "macro not defined" errors).
				// This is conveyed to the parser by returning a different token kind
				// (`T_LABEL` instead of `T_ID`), so that it can immediately dispatch to
				// the code that switches the lexer to "raw" mode.
				//
				// The traditional condition was "does the identifier start with the first character
				// of its line?", but this caused a lot of confusion when people expected to
				// be able to use macros unindented (like instructions and directives can),
				// so the condition has been changed.
```

Reply:
:grimacing: I think that takes it from "paragraph-long exclamation-mark-worthy warning/clarification we should have had to begin with" to "yikes wtf is going on here". It doesn't need to explain related-but-noncentral details like "how do the parser rules for labels and macros work" (go look at them in parser.y) or "historical trivia about how the lexer used to work" (maybe worth a write up *somewhere* but not sandwiched inside live code) "what's lookahead" (please RTFM before touching lexer/parser code).

Reply:
Maybe adapt this comment to go in the relevant test case file?

Reply:
I think the parsing rules explanation is warranted, because this lexer hack *coordinates* with them. I agree to removing the historical section, but again, due to the "coordination" aspect, I think that talking about how *Bison* deals with lookahead is also warranted.

I know parser generators that *always* use lookahead, even when there is no ambiguity. This kind of comment would help if someone decides to compile against $BISON_REPLACEMENT, and finds that things go nuclear. Or, say, [someone looking to reimplement RGBDS](https://github.com/ISSOtm/rsgbds).